### ISSUE TITLE
CDI.current().getBeanManager() causes NullPointerException

### ISSUE THREAD

**Issue by eweitlaner** [OpenLiberty/open-liberty#4218]
Trying to use the JSF2.3 async Websocket Feature in 18.0.0.2. When returning to the SessionScoped Handler and try to send an update message via PushContextImpl.send("xxxx")

During JSF Lifecycle the BeanManager in  PushContextImpl is ok (FacesContext is there).

When returning from the async processing, there is no FacesContext (ok I think) .. the BeanManager gets fetched from CDI.current().getBeanManager() -> NPE (nok)

We tried some other async examples, same issue there. 

Stack:
java.lang.NullPointerException
at org.apache.myfaces.push.cdi.PushContextImpl.send(PushContextImpl.java:71)
at at.gv.tirol.service.web.handler.ServiceHandler.writeServiceResults(ServiceHandler.java:209)
at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1595)
at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

**Comment by benjamin-confino** [comment]
Thanks for the sample. 

The issue is that CDI.current() needs to know which application is calling it so it can return the right CDI. Unfortunately it identifies the application by looking at a ThreadLocal (specifically one inside ComponentMetaDataAccessorImpl) which is not being set on the new thread. 

I believe the correct way to handle this issue is to use a ManagedExecutorService which will ensure the threads are all populated corectly. Unfortunately since the async code is in javax.faces that probably isn't an option. 

I've been asking around to look for a solution, no answers have come up but I'll continue to look. 

Regards
Benjamin

**Comment by eweitlaner** [comment]
We are doing a POC with different AppServers, the async+websocket feature in JSF 2.3 is important to us. We are also WebSphere traditional customers, let me know if we can support you to quickly solve/fix this.

**Comment by c-koell** [comment]
Hi !

We tried with the ManagedExecutorService and then it works fine. But of course it should work with the default Method.

```
@Resource
ManagedExecutorService threadPool;

public void fireAsyncEvent_2() {
    event.fireAsync(new Message("Hello"), NotificationOptions.ofExecutor(threadPool));
}
```

greets
claus

**Comment by benjamin-confino** [comment]
I agree that the default method should work. I've found a potential way to make it work by patching weld and sent them an email to ask about it.

**Comment by benjamin-confino** [comment]
So here's where it stands at the moment. 

I got the ManagedExecutorService working (the bnd needed an update). 

However when Weld tries to use the the ManagedExecutorService during an application startup it cannot handle JndiHelperComponentMetaData as that class doesn't implement IdentifiableComponentMetaData.

IdentifiableComponentMetaData in turn requires a DeferredMetaDataFactory, I tested quickly pointing to the EJB and WEB DeferredMetaDataFactory, but neither worked. We'll probably have to write our own one.

**Comment by benjamin-confino** [comment]
The fix is merged

**Comment by benjamin-confino** [comment]
@c-koell Did you get a chance to test the fix?

**Comment by c-koell** [comment]
Is it included in the 2018.8.0.0 Beta ?
